lib/Agrammon/OutputFormatter/Text.pm6

```diff
@@ -1,73 +1,79 @@
-                            render-filters(@module-lines, $value, $unit, $indent, :$all-filters);
-                        }
+                my $unit = $model.output-unit($module, $output, $language);
+                if $unit {
+                    push @output-lines, "    $output = $val $unit";
+                }
+                else {
+                    push @output-lines, "    $output = $val";
+                }
+                if $include-filters {
+                    if $value ~~ Agrammon::Outputs::FilterGroupCollection && $value.has-filters {
+                        add-filters(@output-lines, $value, $unit, $indent, :$all-filters);
                     }
                 }
             }
+            @lines.append: |@title-lines, |@output-lines if @output-lines;
         }
         when Array {
+            my @subtitle-lines;
+            @output-lines = [];
             for sorted-kv($_) -> $instance-id, %instance-outputs {
                 for sorted-kv(%instance-outputs) -> $fq-name, %values {
                     my $q-name = $module ~ '[' ~ $instance-id ~ ']' ~ $fq-name.substr($module.chars);
-                    push @module-lines, "    $q-name";
+                    push @subtitle-lines, "    $q-name";
                     for sorted-kv(%values) -> $output, $value {
+                        next unless $model.should-print($fq-name, $output, @print-set);
+
                         my $val = flat-value($value // 'UNDEFINED');
-                        my $var-print = $model.output-print($module, $output) ~ ',All';
-                        if not defined $reports-selected or $var-print.split(',') ∩ @print-set {
-                            $n++;
-                            my $unit = $model.output-unit($module, $output, $language);
-                            push @module-lines, "        $output = $val $unit";
-                            if $include-filters {
-                                if $value ~~ Agrammon::Outputs::FilterGroupCollection && $value.has-filters {
-                                    render-filters(@module-lines, $value, $unit, $indent, :$all-filters);
-                                }
+                        my $unit = $model.output-unit($module, $output, $language);
+                        if $unit {
+                            push @output-lines, "        $output = $val $unit";
+                        }
+                        else {
+                            push @output-lines, "        $output = $val";
+                        }
+                        if $include-filters {
+                            if $value ~~ Agrammon::Outputs::FilterGroupCollection && $value.has-filters {
+                                add-filters(@output-lines, $value, $unit, $indent, :$all-filters);
                             }
                         }
                     }
+                    NEXT {
+                        @lines.append: |@title-lines, |@subtitle-lines, |@output-lines if @output-lines;
```

Comment:
I personally find this more understandable than the `$n` thing. :+1:

Reply:
Then I probably will also when I look at it in the future. I must admit that I had to think about the `$n` version a while also to come up with the explanation. :-)